# mechanically kill people to find the answer

In [1]:
def josephmech(n):
    lis = list(range(1,n+1))
    while len(lis)>1:
        if len(lis)%2 == 0:
            lis = lis[::2] #omitting every second number
    #         print(lis) #to check the answer
        elif len(lis)%2 ==1:
            last = lis[-1] #last one won't be killed
            lis = lis[:-2:2]
            lis.insert(0,last) # adding the last to the start
    #         print(lis) #to check the answer
    return lis[0]

In [2]:
%%timeit
(josephmech(10000))

602 µs ± 34.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


# use powers of 2, to do it a little faster

In [3]:
def joseph_use_power_2(n):
    k = 0
    while 2**(k+1)<n:
        k+=1
    if n-(2**(k+1)): #case where m is not equal to zero, i.e, n is not a perfect power of 2
        return (2*(n-(2**k))+1)
    else: #m = 0, n is perfect power
        return 1

In [4]:
%%timeit
(joseph_use_power_2(10000))

13.1 µs ± 688 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


# solution using binary, right shift

In [5]:
def josephbin(n):
    bin_n = bin(n)[2:]
    ans = bin_n[1:]+'1'
    ans = int(ans,base=2)
    return ans

In [6]:
%%timeit
(josephbin(10000))

1.79 µs ± 56.6 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


# case where you are not just killing alternate person but every kth person

here, recrsive solution is needed. by kth person I mean if it starts from 1, k will be killed first then 2k....  
this is just the same like actual josephus problem when k=2

recursion is  
$solution(n, k) = (solution(n - 1, k) + k-1)$%$n + 1 $    
because after 1 person is killed, total people are n-1, and now the starting person is one after the killed person is supposed to be k+1, but in the code it is from 1, to correct that add k-1, and we take remainder after dividing by n because this should be circular. and one because it is actually the next person


and base case is  
$solution(1, k) = 1$

In [7]:
def recur_joseph(n, k): 
    if (n == 1): 
        return 1
    else: 
        return (recur_joseph(n-1,k)+ k-1)% n + 1

## check recursive solution for k=2, actual josephus problem

In [8]:
print(recur_joseph(8,2),recur_joseph(64,2),recur_joseph(50,2),recur_joseph(100,2),recur_joseph(108,2))
# so it is working

1 1 37 73 89


In [16]:
import sys
sys.setrecursionlimit(15000) #this takes so much recursion

In [ ]:
%%timeit
(recur_joseph(10000,2))

## time taken by different methods

- mechanical method -> 602 µs ± 34.2 µs
- using powers of 2 -> 13.1 µs ± 688 ns 
- binary method -> 1.79 µs ± 56.6 ns
- recursive -> is not good for big numbers, exceeds recusrion limit and kernel dies most of the times